File: 0743-network-delay-time/0743-network-delay-time.java
import java.util.*;

class Solution {

    public int networkDelayTime(int[][] times, int n, int k) {
        // Create adjacency list representation of the graph
        Map<Integer, Map<Integer, Integer>> adjList = new HashMap<>();
        for (int i = 1; i <= n; i++) {
            adjList.put(i, new HashMap<>());
        }
        for (int[] arr : times) {
            adjList.get(arr[0]).put(arr[1], arr[2]);
        }

        // Initialize distances map with infinity for all nodes except the starting node k
        Map<Integer, Integer> distances = new HashMap<>();

        for (int i=1; i<=n; i++){
            distances.put(i, Integer.MAX_VALUE);
        }
        
        distances.put(k, 0);

        // Priority queue to keep track of nodes to process, starting with node k
        PriorityQueue<Map.Entry<Integer, Integer>> pq = new PriorityQueue<>(Map.Entry.comparingByValue());
        pq.offer(Map.entry(k, 0));

        // Main loop to process nodes in the priority queue
        while (!pq.isEmpty()) {
            Map.Entry<Integer, Integer> current = pq.poll();
            int currentNode = current.getKey();
            int currentDist = current.getValue();

            // Skip if a shorter distance to the current node is already known
            if (currentDist > distances.get(currentNode)) {
                continue;
            }

            // Iterate through all neighbors of the current node
            for (Map.Entry<Integer, Integer> neighbor : adjList.get(currentNode).entrySet()) {
                int newDist = currentDist + neighbor.getValue();
                // If a shorter path to the neighbor is found, update and add to the queue
                if (newDist < distances.get(neighbor.getKey())) {
                    distances.put(neighbor.getKey(), newDist);
                    pq.offer(Map.entry(neighbor.getKey(), newDist));
                }
            }
        }

        // Find the maximum distance from the starting node to any other node
        int ans = Integer.MIN_VALUE;
        for (int distance : distances.values()) {
            if (distance == Integer.MAX_VALUE) return -1; // If any node is unreachable
            ans = Math.max(ans, distance);
        }
        return ans;
    }
}

File: 0743-network-delay-time/0743-network-delay-time.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {

        // {node: {node, distance}}
        unordered_map<int, unordered_map<int, int>> adj;

        for(const auto &t: times){
            adj[t[0]][t[1]] = t[2];
        }
        // {node, distance}
        unordered_map<int, int> distances;

        for(int i=1; i<=n; i++){
            distances[i] = INT_MAX;
        }

        distances[k] = 0;

        // {distance, node} min heap
        priority_queue<pair<int, int>, vector<pair<int,int>>, greater<pair<int, int>>> pq;

        pq.push({0, k});

        while(!pq.empty()){

            int currNode = pq.top().second;
            int currDist = pq.top().first;
            pq.pop();

            if(currDist > distances[currNode]) continue;

            for(const auto &neighbor: adj[currNode]){
                int newDist = currDist + neighbor.second;

                if(newDist < distances[neighbor.first]){
                    distances[neighbor.first] = newDist;
                    pq.push({newDist, neighbor.first});
                }
            }
        }
        int ans = INT_MIN;
        for(const auto &[node, distance]: distances){
            if(distance == INT_MAX) return -1;
            ans = max(ans, distance);
        }

        return ans;
    }
};